### PR TITLE
operator: Provide community operator bundle

### PR DESCRIPTION
**Pull request by periklis**
**What this PR does / why we need it**:
The Loki Operator is currently for the community users only available via our makefile targets. In preparation of a future release on operatorhub.io, this change set adds support for building two bundles in parallel one for the community and one for OpenShift. This enables community users to use OLM on Kubernetes.

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Special notes for your reviewer**:

**Checklist**
- [x] Reviewed the `CONTRIBUTING.md` guide
- [ ] Documentation added
- [ ] Tests updated
- [ ] `CHANGELOG.md` updated
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

### CODE REVIEW

**Review by Red-GV — COMMENTED** [review]
Shouldn't the `properties.yaml` file be moved to `bundle/openshift/metadata`?

**Review by Red-GV — COMMENTED** [review]
Do we need two scorecard tests or just one for the GitHub? I think the location of the file for the GitHub test has changed.

**Review by Red-GV — COMMENTED** [review]
Overall, looks fine to me. Just a small nitpick.

File: operator/config/manifests/community/bases/loki-operator.clusterserviceversion.yaml [hunk 1]
```diff
@@ -0,0 +1,1477 @@
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+metadata:
+  annotations:
+    alm-examples: '[]'
+    capabilities: Full Lifecycle
```
Comment (xperimental) [inline operator/config/manifests/community/bases/loki-operator.clusterserviceversion.yaml:6]: I wondered a few times already if "Full Lifecycle" might be a bit too much for now. Should we lower this to the next lower one (which afaik is "Seamless Upgrades").

**Comment by xperimental** [comment]
Sorry, was too fast clicking "submit":

While trying to use the usual `make` targets I wondered if it was intended that `olm-deploy` and related targets work with the community version of the operator as well. The Makefile seems a bit "undecided" if it should use extra variables for a few things and then overwrite a few other variables for the community-variant. I think I'd prefer it if the "normal variables" would be used for the community-variant as well, only overridden with different defaults. That should make it easier to re-use the same targets for the community version.

File: operator/Makefile [hunk 2]
```diff
@@ -26,9 +26,12 @@ LOKI_OPERATOR_NS ?= openshift-operators-redhat
 # To re-generate a bundle for another specific version without changing the standard setup, you can:
 # - use the VERSION as arg of the bundle target (e.g make bundle VERSION=0.0.2)
 # - use environment variables to overwrite this value (e.g export VERSION=0.0.2)
-VERSION ?= v0.0.1
+VERSION ?= v0.1.0
+COMMUNITY_VERSION ?= main-39f2856
+COMMUNITY_RELEASE_VERSION=0.1.0
 CHANNELS ?= "stable"
 DEFAULT_CHANNEL ?= "stable"
+SUPPORTED_OCP_VERSIONS="v4.10"
```
[the remark this answers is not in the document]
Comment (periklis) [reply operator/Makefile:34]: According to our official public [docs](https://redhat-connect.gitbook.io/certified-operator-guide/ocp-deployment/operator-metadata/bundle-directory/managing-openshift-versions) this is the start of an open range. It translates to `starting with 4.10 and anything above`.

File: operator/Makefile [hunk 3]
```diff
@@ -80,6 +84,16 @@ else
 GOBIN=$(shell go env GOBIN)
 endif
 
+USE_COMMUNITY_BUILD ?= false
+ifeq ($(USE_COMMUNITY_BUILD), true)
```
[the remark this answers is not in the document]
Comment (periklis) [reply operator/Makefile:88]: Something we should aim for later after releasing the first v1.0.0 community operator.

File: operator/bundle/community/manifests/loki-operator-controller-manager-metrics-service_v1_service.yaml [hunk 4]
```diff
@@ -0,0 +1,24 @@
+apiVersion: v1
+kind: Service
+metadata:
+  creationTimestamp: null
+  labels:
+    app.kubernetes.io/instance: loki-operator-v0.0.1
```
[the remark this answers is not in the document]
Comment (periklis) [reply operator/bundle/community/manifests/loki-operator-controller-manager-metrics-service_v1_service.yaml:6]: Good catch! it should imho.

File: operator/config/manifests/community/bases/loki-operator.clusterserviceversion.yaml (hunk 1 above)
Comment (periklis) [reply operator/config/manifests/community/bases/loki-operator.clusterserviceversion.yaml:6]: Revisiting the [capability levels](https://sdk.operatorframework.io/docs/overview/operator-capabilities/#level-3---full-lifecycle) I still believe that the Loki Operator manages the full-lifecycle of loki. Couple of them don't really match to Loki (e.g. backup/restore), but we apply liveness/readiness probes, CPU/Memory requests, rolling restart strategy to Loki and the other operands. The only thing we miss is PodDisruptionBudget, but this is neglible for `1x.small` as it has only 2 replicas per component and might be only a thing to consider later for `1x.medium`'s ingesters and queriers.

File: operator/hack/operatorhub.sh [hunk 5]
```diff
@@ -0,0 +1,67 @@
+#!/bin/bash
+
+COMMUNITY_OPERATORS_REPOSITORY="k8s-operatorhub/community-operators"
+UPSTREAM_REPOSITORY="redhat-openshift-ecosystem/community-operators-prod"
+LOCAL_REPOSITORIES_PATH=${LOCAL_REPOSITORIES_PATH:-"$(dirname $(dirname $(dirname $(pwd))))"}
```
[the remark this answers is not in the document]
Comment (periklis) [reply operator/hack/operatorhub.sh:5]: The thing here is that based on the local repository path I want to figure out the sibling repositories. How does the command you suggest help here? I failed to get this running or applying.

**Comment by periklis** [comment]
> Sorry, was too fast clicking "submit":
> 
> While trying to use the usual `make` targets I wondered if it was intended that `olm-deploy` and related targets work with the community version of the operator as well. The Makefile seems a bit "undecided" if it should use extra variables for a few things and then overwrite a few other variables for the community-variant. I think I'd prefer it if the "normal variables" would be used for the community-variant as well, only overridden with different defaults. That should make it easier to re-use the same targets for the community version.

Actually the `make {olm-deploy|olm-upgrade|olm-undeploy}` commands work the same with the community bundle as with the openshift bundle like this, e.g.:
```
make olm-deploy REGISTRY_ORG_COMMUNITY=theperiklis USE_COMMUNITY_BUILD=true
```
Currently using the community bundle is an opt-in procedure with `USE_COMMUNITY_BUILD=true`, but we should aim to flip this towards a Community Loki Operator v1.0.0 release. My intention here is to keep dev experience changes to a minimum for now.

File: operator/hack/operatorhub.sh [hunk 6]
```diff
@@ -0,0 +1,69 @@
+#!/usr/bin/env bash
```
Comment (xperimental) [inline operator/hack/operatorhub.sh:1]: Does this shell script need the usual `set -e -u -o pipefail`?

File: operator/bundle/community/manifests/loki-operator-manager-config_v1_configmap.yaml [hunk 7]
```diff
@@ -0,0 +1,32 @@
+apiVersion: v1
+data:
+  controller_manager_config.yaml: |
+    apiVersion: config.loki.grafana.com/v1
+    kind: ProjectConfig
+    health:
+      healthProbeBindAddress: :8081
+    metrics:
+      bindAddress: :8080
```
[the remark this answers is not in the document]
Comment (xperimental) [reply operator/bundle/community/manifests/loki-operator-manager-config_v1_configmap.yaml:9]: The above comment might have gotten lost in the Github UI...

**Review by xperimental — COMMENTED** [review]
Looks good from my tests. Just two small nitpicks left.

File: operator/bundle/openshift/metadata/properties.yaml [hunk 8]
```diff
@@ -1,3 +1,3 @@
 properties:
   - type: olm.maxOpenShiftVersion
-    value: 4.15
+    value: 4.12
```
Comment (xperimental) [inline operator/bundle/openshift/metadata/properties.yaml:3]: This seems to be an unwanted downgrade.

Comment (periklis) [reply operator/bundle/openshift/metadata/properties.yaml]: Good catch rebase error by one ;)

**Review by xperimental — APPROVED** [review]

**Pull request merged**